core/state/statedb.go

```diff
@@ -1102,3 +1127,13 @@ func (s *StateDB) AddressInAccessList(addr common.Address) bool {
 func (s *StateDB) SlotInAccessList(addr common.Address, slot common.Hash) (addressPresent bool, slotPresent bool) {
 	return s.accessList.Contains(addr, slot)
 }
+
+// GetPredicateStorageSlots returns the storage slots associated with a given address, and whether or not
+// that address was included in the optional access list of the transaction.
+// The storage slots are returned in the same order as they appeared in the transaction.
+// These are the same storage slots that are used to verify any transaction
+// predicates for transactions with access list addresses that match a precompile address.
+func (s *StateDB) GetPredicateStorageSlots(address common.Address) ([]byte, bool) {
```

Comment:
I think you called this out someplace else. But could `GetPredicateStorageSlots` just return a `[]byte` and `nil` if the address wasn't included in the access list?

I guess the only behavior difference is that the case when the address was not included at all and the case when the address was included with an empty storage slots array would be indistinguishable. I'm not sure whether there is a use case for a precompile predicate to need to be invoked with `nil` data, but probably best to leave in the option for that either way, so I'd be in favor of keeping this as is.

Reply:
Good callout. I agree it would be useful to use `exists` to track whether the precompile address was included and make it so that you can distinguish between including an empty array and not including it at all.

We should also make sure to do a pass over all of invariants to ensure:
1. They are clearly commented in the code
2. We are all on the same page and understand them
3. Write clear documentation for them